cmd/backup.go

```diff
@@ -0,0 +1,139 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/spf13/cobra"
+	"github.com/tigrisdata/tigris-cli/client"
+	"github.com/tigrisdata/tigris-cli/util"
+	"github.com/tigrisdata/tigris-client-go/driver"
+)
+
+func listDatabases(ctx context.Context, dbFilter []string) ([]string, error) {
+	databases := []string{}
+
+	resp, err := client.Get().ListDatabases(ctx)
+	if err != nil {
+		return databases, util.Error(err, "list databases")
+	}
+
+	for _, v := range resp {
+		if len(dbFilter) == 0 || util.Contains(dbFilter, v) {
+			databases = append(databases, v)
+		}
+	}
+
+	return databases, nil
+}
+
+func listCollections(ctx context.Context, dbName string, collectionFilter []string) ([]string, error) {
+	collections := []string{}
+
+	resp, err := client.Get().UseDatabase(dbName).ListCollections(ctx)
+	if err != nil {
+		return collections, util.Error(err, "list collections")
+	}
+
+	for _, v := range resp {
+		if len(collectionFilter) == 0 || util.Contains(collectionFilter, v) {
+			collections = append(collections, v)
+		}
+	}
+
+	return collections, nil
+}
+
+var backupCmd = &cobra.Command{
+	Use:   "backup [filters]",
+	Short: "Dumps documents and schemas to JSON files",
+	Long: `Dumps documents and schemas to JSON files
+	in the current working directory.
+	
+	If a database filter is provided it only dumps the schemas of the databases specified.
+	Likewise, collection filters will limit the output to matching collection names.`,
+	Args: cobra.MinimumNArgs(0),
+	Run: func(cmd *cobra.Command, args []string) {
+		withLogin(cmd.Context(), func(ctx context.Context) error {
+			dbFilter := []string{}
+			collectionFilter := []string{}
+
+			ctx, cancel := context.WithTimeout(ctx, 1*time.Hour)
+			defer cancel()
+			os.Setenv("TIGRIS_TIMEOUT", "1h")
+
+			databases, err := listDatabases(ctx, dbFilter)
+			if err != nil {
+				util.Stderrf("%v\n", err)
+			}
+
+			for _, db := range databases {
+				collections, err := listCollections(ctx, db, collectionFilter)
+				if err != nil {
+					return util.Error(err, "error listing collections")
+				}
+
+				// Dump schema
+				resp, err := client.Get().DescribeDatabase(ctx, db)
+				if err != nil {
+					return util.Error(err, "describe collection failed")
+				}
+
+				filename := fmt.Sprintf("./%s.schema", db)
+				for _, v := range resp.Collections {
+					err := os.WriteFile(filename, []byte(string(v.Schema)), 0o600)
```

Comment:
Return util.Error only after Tigris API call which can return `Unauthenticated` error.

Otherwise with should use Fatal:
```
err := os.WriteFile(filename, []byte(string(v.Schema)), 0o600)
util.Fatal(err, "writing schema file")
```

Reply:
Makes sense. Thanks!